src/loki/LokiBase.cpp

```diff
@@ -0,0 +1,256 @@
+  Stats.create("receive.packets", Counters.RxPackets);
+  Stats.create("receive.bytes", Counters.RxBytes);
+  Stats.create("receive.dropped", Counters.FifoPushErrors);
+  Stats.create("receive.fifo_seq_errors", Counters.FifoSeqErrors);
+
+  Stats.create("readouts.count", Counters.ReadoutsCount);
+  Stats.create("readouts.error_bytes", Counters.ReadoutsErrorBytes);
+
+  Stats.create("thread.processing_idle", Counters.RxIdle);
+
+  Stats.create("events.count", Counters.Events);
+  Stats.create("events.udder", Counters.EventsUdder);
+  Stats.create("events.geometry_errors", Counters.GeometryErrors);
+
+  Stats.create("transmit.bytes", Counters.TxBytes);
+
+  /// \todo below stats are common to all detectors and could/should be moved
+  Stats.create("kafka.produce_fails", Counters.kafka_produce_fails);
+  Stats.create("kafka.ev_errors", Counters.kafka_ev_errors);
+  Stats.create("kafka.ev_others", Counters.kafka_ev_others);
+  Stats.create("kafka.dr_errors", Counters.kafka_dr_errors);
+  Stats.create("kafka.dr_others", Counters.kafka_dr_noerrors);
+  // clang-format on
+
+  std::function<void()> inputFunc = [this]() { LokiBase::input_thread(); };
+  Detector::AddThreadFunction(inputFunc, "input");
+
+  std::function<void()> processingFunc = [this]() {
+    LokiBase::processing_thread();
+  };
+  Detector::AddThreadFunction(processingFunc, "processing");
+
+  XTRACE(INIT, ALW, "Creating %d Loki Rx ringbuffers of size %d",
+         EthernetBufferMaxEntries, EthernetBufferSize);
+  /// \todo the number 11 is a workaround
+  EthernetRingbuffer = new RingBuffer<EthernetBufferSize>(EthernetBufferMaxEntries + 11);
+  assert(EthernetRingbuffer != 0);
+}
+
+void LokiBase::input_thread() {
+  /** Connection setup */
+  Socket::Endpoint local(EFUSettings.DetectorAddress.c_str(),
+                         EFUSettings.DetectorPort);
+  UDPReceiver receiver(local);
+  // receiver.buflen(opts->buflen);
+  receiver.setBufferSizes(0, EFUSettings.DetectorRxBufferSize);
+  receiver.printBufferSizes();
+  receiver.setRecvTimeout(0, 100000); /// secs, usecs 1/10s
+
+  for (;;) {
+    int rdsize;
+    unsigned int eth_index = EthernetRingbuffer->getDataIndex();
+
+    /** this is the processing step */
+    EthernetRingbuffer->setDataLength(eth_index, 0);
+    if ((rdsize = receiver.receive(EthernetRingbuffer->getDataBuffer(eth_index),
+                                   EthernetRingbuffer->getMaxBufSize())) > 0) {
+      EthernetRingbuffer->setDataLength(eth_index, rdsize);
+      XTRACE(INPUT, DEB, "Received an udp packet of length %d bytes", rdsize);
+      Counters.RxPackets++;
+      Counters.RxBytes += rdsize;
+
+      if (InputFifo.push(eth_index) == false) {
+        Counters.FifoPushErrors++;
+      } else {
+        EthernetRingbuffer->getNextBuffer();
+      }
+    }
+
+    // Checking for exit
+    if (not runThreads) {
+      XTRACE(INPUT, ALW, "Stopping input thread.");
+      return;
+    }
+  }
+}
+
+void LokiBase::processing_thread() {
```

Comment:
This function is too long. It should be split into two. This will aid in understanding that there are actually two separate code paths here.